File: app/src/main/java/com/airadar/app/data/OpenSkyClient.kt
package com.airadar.app.data

import com.airadar.app.BuildConfig
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import okhttp3.FormBody
import okhttp3.OkHttpClient
import okhttp3.Request
import org.json.JSONArray
import org.json.JSONObject
import java.io.IOException
import java.time.Duration
import java.time.Instant
import java.time.LocalDate
import java.util.concurrent.TimeUnit

/**
 * Pulls the positions an aircraft actually reported over ADS-B, so a flown leg can
 * be drawn as the path it took rather than a great circle.
 *
 * OpenSky keys everything by the airframe's ICAO24 transponder hex, not by flight
 * number, so resolving a leg is two steps: list departures from the origin around
 * the scheduled time and pick the one whose ATC callsign matches, then ask for
 * that airframe's track at a moment during the flight.
 */
object OpenSkyClient {

    private const val TOKEN_URL =
        "https://auth.opensky-network.org/auth/realms/opensky-network/protocol/openid-connect/token"
    private const val API = "https://opensky-network.org/api"

    private val http = OkHttpClient.Builder()
        .connectTimeout(15, TimeUnit.SECONDS)
        .readTimeout(30, TimeUnit.SECONDS)
        .build()

    private var token: String? = null
    private var tokenExpiresAt: Instant = Instant.EPOCH

    class OpenSkyException(message: String) : IOException(message)

    /** Positions plus the day they were flown — an earlier day when borrowed for a future leg. */
    data class FetchedTrack(val points: List<TrackPoint>, val flownOn: LocalDate)

    suspend fun fetchTrack(flight: Flight): FetchedTrack = withContext(Dispatchers.IO) {
        val callsign = flight.callsign
            ?: throw OpenSkyException("No ATC callsign known for ${flight.flightNumber}")
        val origin = flight.departureAirport
            ?: throw OpenSkyException("Unknown departure airport ${flight.departure}")
        val destination = flight.arrivalAirport
            ?: throw OpenSkyException("Unknown arrival airport ${flight.arrival}")
        val scheduled = flight.departureInstant
            ?: throw OpenSkyException("No departure time for ${flight.flightNumber}")

        val bearer = accessToken()

        when (flight.phase) {
            // Airborne now: the flights endpoints only list finished flights, and
            // adsb.lol's live lookup (LivePositionClient) resolves the callsign to
            // its hex far more simply than OpenSky's own bounding-box scan of every
            // live state vector did.
            FlightPhase.IN_PROGRESS -> {
                val hex = LivePositionClient.position(callsign)?.hex
                    ?: throw OpenSkyException(
                        "$callsign is not in the live picture right now. Either no " +
                                "receiver can hear it, or it is not actually in the air."
                    )
                // Only this flight's own fixes: until OpenSky picks it up again
                // after take-off, its latest track is the leg that brought it in.
                val start = flight.trackStart ?: scheduled
                val points = fetchPath(bearer, hex, at = 0).filter { p -> p.time?.let { !it.isBefore(start) } ?: true }
                if (points.size < 2) throw OpenSkyException(
                    "OpenSky hasn't picked up ${flight.flightNumber} since take-off yet; its latest track is the aircraft's previous flight."
                )
                return@withContext FetchedTrack(points, Instant.now().atZone(origin.zone).toLocalDate())
            }

            // Flown: its own day. Future: the most recent day this callsign left the
            // same airport - airlines fly the same routing day after day.
            FlightPhase.PAST, FlightPhase.UPCOMING -> {
                val days: List<Instant> = if (flight.phase == FlightPhase.PAST) listOf(scheduled)
                else (1..7L).map { minOf(scheduled, Instant.now()).minus(Duration.ofDays(it)) }

                val seenFromAirline = linkedSetOf<String>()
                val checked = mutableListOf<String>()
                for (day in days) {
                    checked += day.atZone(origin.zone).toLocalDate().toString().substring(5)
                    val found = findFlight(
                        bearer, callsign, origin, destination, day,
                        flight.durationMinutes, seenFromAirline
                    ) ?: continue
                    val (icao24, firstSeen, lastSeen) = found
                    val flownOn = Instant.ofEpochSecond(firstSeen).atZone(origin.zone).toLocalDate()
                    return@withContext FetchedTrack(
                        fetchPath(bearer, icao24, at = (firstSeen + lastSeen) / 2),
                        flownOn
                    )
                }

                val where = "${origin.iata} departures, ${destination.iata} arrivals and the global list"
                val hint = if (seenFromAirline.isEmpty())
                    " No ${callsign.takeWhile { it.isLetter() }} flight at all was seen there."
                else " Same-airline callsigns it did see: ${seenFromAirline.take(6).joinToString()}."
                throw OpenSkyException(
                    if (flight.phase == FlightPhase.UPCOMING)
                        "OpenSky has no $callsign in $where on ${checked.first()}–${checked.last()}." +
                                hint + " Coverage relies on volunteer receivers."
                    else
                        "OpenSky has no $callsign in $where on ${checked.first()}." +
                                hint + " Its history only reaches back about 30 days."
                )
            }
        }
    }

    /**
     * Resolves a finished flight to its airframe. Three lists, cheapest and most
     * specific first: departures at the origin, arrivals at the destination (a flight
     * out of a thinly covered airport is often only caught landing somewhere dense),
     * then the global list in two-hour slices for the cases where OpenSky could not
     * guess either airport.
     */
    private fun findFlight(
        bearer: String,
        callsign: String,
        origin: Airport,
        destination: Airport,
        departAround: Instant,
        durationMinutes: Int,
        seenFromAirline: MutableSet<String>
    ): Triple<String, Long, Long>? {
        val prefix = callsign.takeWhile { it.isLetter() }

        fun pick(body: String?): Triple<String, Long, Long>? {
            val entries = body?.let { JSONArray(it).asObjects() } ?: return null
            entries.map { it.optString("callsign").trim() }
                .filter { it.startsWith(prefix) && it.isNotEmpty() }
                .forEach(seenFromAirline::add)
            // OpenSky's estimated airports are guesses; callsign plus the list we
            // asked for already pins the flight down.
            val match = entries.firstOrNull { sameCallsign(it.optString("callsign"), callsign) }
                ?: return null
            return Triple(match.getString("icao24"), match.getLong("firstSeen"), match.getLong("lastSeen"))
        }

        // Departures drift from schedule; -3h/+6h covers a long delay.
        val begin = departAround.minus(Duration.ofHours(3)).epochSecond
        val end = departAround.plus(Duration.ofHours(6)).epochSecond
        pick(getJson("$API/flights/departure?airport=${origin.icao}&begin=$begin&end=$end", bearer))
            ?.let { return it }

        // Arrivals land a flight-time later; stretch the window by the block time.
        val arrivalEnd = end + durationMinutes * 60L + 3600
        pick(getJson("$API/flights/arrival?airport=${destination.icao}&begin=$begin&end=$arrivalEnd", bearer))
            ?.let { return it }

        // Global fallback, capped at 2 h per call: the slice around schedule, then the
        // one after it for a late departure.
        for (offsetHours in listOf(-1L, 1L)) {
            val b = departAround.plus(Duration.ofHours(offsetHours)).epochSecond
            val e = b + 2 * 3600
            pick(getJson("$API/flights/all?begin=$b&end=$e", bearer))?.let { return it }
        }
        return null
    }

    /** The current live state for one already-known airframe -- cheap (a
     * single icao24 filter), meant to run alongside adsb.lol every poll:
     * each source misses a real fraction of the time (confirmed live -- an
     * aircraft adsb.lol was seeing seconds earlier came back `states: null`
     * here), so whichever one answers on a given cycle covers for the
     * other. null for any failure (network, auth, or genuinely no coverage
     * right now) -- always best-effort, never thrown, since a missed poll
     * from one source is routine, not an error. */
    suspend fun liveState(icao24: String): LivePosition? = withContext(Dispatchers.IO) {
        val bearer = runCatching { accessToken() }.getOrNull() ?: return@withContext null
        val body = runCatching { getJson("$API/states/all?icao24=${icao24.lowercase()}", bearer) }.getOrNull()
            ?: return@withContext null
        val states = runCatching { JSONObject(body).optJSONArray("states") }.getOrNull() ?: return@withContext null
        if (states.length() == 0) return@withContext null
        val v = states.getJSONArray(0)
        if (v.length() <= 10 || v.isNull(5) || v.isNull(6)) return@withContext null
        val heading = if (v.isNull(10)) 0.0 else v.getDouble(10)
        val altMeters = if (v.isNull(7)) null else v.getDouble(7)
        val lastContact = if (v.isNull(4)) null else v.getLong(4)
        LivePosition(
            lat = v.getDouble(6), lon = v.getDouble(5), heading = heading,
            altitudeFeet = altMeters?.let { (it * 3.28084).toInt() }, hex = icao24,
            seenAt = lastContact?.let { Instant.ofEpochSecond(it) } ?: Instant.now()
        )
    }

    /** A first fix for a callsign whose hex isn't known yet -- OpenSky's own
     * live state scanned inside the route's bounding box, the same way the
     * old (removed) bounding-box scan used to work for every poll. More
     * expensive than the icao24-filtered lookup above, so only meant to run
     * once, to learn the hex -- adsb.lol's own direct callsign lookup gets
     * first try, this is only the fallback when that one comes up empty. */
    suspend fun liveState(callsign: String, origin: Airport, destination: Airport): LivePosition? =
        withContext(Dispatchers.IO) {
            val bearer = runCatching { accessToken() }.getOrNull() ?: return@withContext null
            val pad = 4.0
            val lamin = minOf(origin.latitude, destination.latitude) - pad
            val lamax = maxOf(origin.latitude, destination.latitude) + pad
            val lomin = minOf(origin.longitude, destination.longitude) - pad
            val lomax = maxOf(origin.longitude, destination.longitude) + pad
            val body = runCatching {
                getJson("$API/states/all?lamin=$lamin&lomin=$lomin&lamax=$lamax&lomax=$lomax", bearer)
            }.getOrNull() ?: return@withContext null
            val states = runCatching { JSONObject(body).optJSONArray("states") }.getOrNull() ?: return@withContext null
            for (i in 0 until states.length()) {
                val v = states.getJSONArray(i)
                if (v.length() <= 10 || v.isNull(0) || v.isNull(1) || v.isNull(5) || v.isNull(6)) continue
                if (!sameCallsign(v.getString(1), callsign)) continue
                val heading = if (v.isNull(10)) 0.0 else v.getDouble(10)
                val altMeters = if (v.isNull(7)) null else v.getDouble(7)
                val lastContact = if (v.isNull(4)) null else v.getLong(4)
                return@withContext LivePosition(
                    lat = v.getDouble(6), lon = v.getDouble(5), heading = heading,
                    altitudeFeet = altMeters?.let { (it * 3.28084).toInt() }, hex = v.getString(0),
                    seenAt = lastContact?.let { Instant.ofEpochSecond(it) } ?: Instant.now()
                )
            }
            null
        }

    /** OpenSky pads callsigns to eight characters and some carriers zero-pad the number. */
    private fun sameCallsign(seen: String, wanted: String): Boolean {
        fun norm(s: String): String {
            val t = s.trim().uppercase()
            val letters = t.takeWhile { it.isLetter() }
            val digits = t.drop(letters.length).takeWhile { it.isDigit() }.trimStart('0')
            return letters + digits
        }
        return norm(seen) == norm(wanted) && norm(wanted).isNotEmpty()
    }

    /** Positions for one airframe; `at = 0` means the track of the flight it is on now. */
    private fun fetchPath(bearer: String, icao24: String, at: Long): List<TrackPoint> {
        val track = getJson("$API/tracks/all?icao24=$icao24&time=$at", bearer)
            ?: throw OpenSkyException(
                "OpenSky matched the aircraft ($icao24) but kept no track for that flight."
            )
        val path = JSONObject(track).optJSONArray("path")
            ?: throw OpenSkyException("OpenSky returned a track with no positions.")

        return (0 until path.length())
            .map { path.getJSONArray(it) }
            .filter { !it.isNull(1) && !it.isNull(2) }
            .map {
                // path[0] is the point's own Unix time.
                val time = if (!it.isNull(0)) Instant.ofEpochSecond(it.getDouble(0).toLong()) else null
                TrackPoint(lat = it.getDouble(1), lon = it.getDouble(2), time = time)
            }
            .also { if (it.size < 2) throw OpenSkyException("Track too short to draw.") }
    }

    /** The same airframe's most recently completed flight before some point in
     * time -- "this plane flew in from X about N hours ago" material. Purely a
     * courtesy note: never treated as a source of truth the way a real flight's
     * own facts are, and it's fine for this to come back null (no ADS-B
     * history, or it was already on the ground for a while beforehand). */
    data class PreviousFlight(val fromICAO: String?, val toICAO: String?, val landedAt: Instant)

    suspend fun previousFlight(icao24: String, before: Instant): PreviousFlight? = withContext(Dispatchers.IO) {
        val bearer = accessToken()
        val begin = before.epochSecond - 18 * 3600
        val end = before.epochSecond
        val body = getJson("$API/flights/aircraft/$icao24?begin=$begin&end=$end", bearer) ?: return@withContext null
        val entries = runCatching { JSONArray(body).asObjects() }.getOrNull() ?: return@withContext null
        if (entries.isEmpty()) return@withContext null
        val last = entries.filter { it.has("lastSeen") && !it.isNull("lastSeen") }
            .maxByOrNull { it.getLong("lastSeen") } ?: return@withContext null
        fun icao(key: String) = last.optString(key).takeIf { last.has(key) && !last.isNull(key) && it.isNotBlank() }
        PreviousFlight(
            fromICAO = icao("estDepartureAirport"),
            toICAO = icao("estArrivalAirport"),
            landedAt = Instant.ofEpochSecond(last.getLong("lastSeen"))
        )
    }

    private fun accessToken(): String {
        token?.let { if (Instant.now().isBefore(tokenExpiresAt)) return it }

        val id = BuildConfig.OPENSKY_CLIENT_ID
        val secret = BuildConfig.OPENSKY_CLIENT_SECRET
        if (id.isBlank() || secret.isBlank()) {
            throw OpenSkyException(
                "OpenSky credentials missing — add opensky.clientId and " +
                        "opensky.clientSecret to local.properties and rebuild."
            )
        }

        val request = Request.Builder()
            .url(TOKEN_URL)
            .post(
                FormBody.Builder()
                    .add("grant_type", "client_credentials")
                    .add("client_id", id)
                    .add("client_secret", secret)
                    .build()
            )
            .build()

        http.newCall(request).execute().use { response ->
            if (response.code == 401 || response.code == 400) {
                throw OpenSkyException("OpenSky rejected the client credentials.")
            }
            if (!response.isSuccessful) {
                throw OpenSkyException("OpenSky sign-in failed (HTTP ${response.code}).")
            }
            val body = JSONObject(response.body?.string().orEmpty())
            val access = body.getString("access_token")
            // Renew a minute early so a call never straddles expiry.
            val ttl = body.optLong("expires_in", 1800) - 60
            token = access
            tokenExpiresAt = Instant.now().plusSeconds(ttl)
            return access
        }
    }

    /** Response body, or null when OpenSky simply has nothing for that query (404). */
    private fun getJson(url: String, bearer: String): String? {
        val request = Request.Builder()
            .url(url)
            .header("Authorization", "Bearer $bearer")
            .build()
        http.newCall(request).execute().use { response ->
            when {
                response.code == 404 -> return null
                response.code == 429 -> throw OpenSkyException(
                    "OpenSky rate limit reached for today. Try again later."
                )
                !response.isSuccessful -> throw OpenSkyException(
                    "OpenSky request failed (HTTP ${response.code})."
                )
            }
            return response.body?.string().orEmpty()
        }
    }

    private fun JSONArray.asObjects(): List<JSONObject> =
        (0 until length()).map { getJSONObject(it) }
}

File: ios/Airadar/Data/OpenSkyClient.swift
import Foundation

/// Positions an aircraft actually reported over ADS-B, so a flown leg is drawn as
/// the path it took. OpenSky keys everything by the airframe's ICAO24 hex, so a leg
/// is resolved in two steps: find the flight by callsign, then ask for its track.
actor OpenSkyClient {
    static let shared = OpenSkyClient()

    private static let tokenURL = URL(string: "https://auth.opensky-network.org/auth/realms/opensky-network/protocol/openid-connect/token")!
    private static let api = "https://opensky-network.org/api"

    struct OpenSkyError: LocalizedError { let message: String; var errorDescription: String? { message } }

    /// Positions plus the day they were flown.
    struct FetchedTrack: Sendable { let points: [TrackPoint]; let flownOn: String }

    private var token: String?
    private var tokenExpiresAt = Date.distantPast

    func fetchTrack(_ flight: Flight) async throws -> FetchedTrack {
        guard let callsign = flight.callsign else { throw OpenSkyError(message: "No ATC callsign known for \(flight.flightNumber)") }
        guard let origin = flight.departureAirport else { throw OpenSkyError(message: "Unknown departure airport \(flight.departure)") }
        guard let destination = flight.arrivalAirport else { throw OpenSkyError(message: "Unknown arrival airport \(flight.arrival)") }
        guard let scheduled = flight.departureInstant else { throw OpenSkyError(message: "No departure time for \(flight.flightNumber)") }
        let bearer = try await accessToken()

        switch flight.phase {
        case .inProgress:
            // Airborne now: the flights endpoints only list finished flights; look at live state vectors.
            guard let icao24 = try await findAirborne(bearer, callsign, origin, destination) else {
                throw OpenSkyError(message: "\(callsign) is not in OpenSky's live picture right now. Either no receiver can hear it, or it is not actually in the air.")
            }
            return FetchedTrack(points: try await fetchPath(bearer, icao24, at: 0),
                                flownOn: LocalDateTime.from(Date(), in: origin.zone).dayString)

        case .past, .upcoming:
            let days: [Date] = flight.phase == .past ? [scheduled]
                : (1...7).map { min(scheduled, Date()).addingTimeInterval(-Double($0) * 86400) }
            var seenFromAirline: [String] = []
            var checked: [String] = []
            for day in days {
                checked.append(String(LocalDateTime.from(day, in: origin.zone).dayString.dropFirst(5)))
                if let (icao24, first, last) = try await findFlight(bearer, callsign, origin, destination, day, flight.durationMinutes, &seenFromAirline) {
                    let flownOn = LocalDateTime.from(Date(timeIntervalSince1970: TimeInterval(first)), in: origin.zone).dayString
                    return FetchedTrack(points: try await fetchPath(bearer, icao24, at: (first + last) / 2), flownOn: flownOn)
                }
            }
            let prefix = String(callsign.prefix { $0.isLetter })
            let hint = seenFromAirline.isEmpty ? " No \(prefix) flight at all was seen there."
                : " Same-airline callsigns it did see: \(seenFromAirline.prefix(6).joined(separator: ", "))."
            let whereText = "\(origin.iata) departures, \(destination.iata) arrivals and the global list"
            throw OpenSkyError(message: flight.phase == .upcoming
                ? "OpenSky has no \(callsign) in \(whereText) on \(checked.first ?? "")–\(checked.last ?? "")." + hint + " Coverage relies on volunteer receivers."
                : "OpenSky has no \(callsign) in \(whereText) on \(checked.first ?? "")." + hint + " Its history only reaches back about 30 days.")
        }
    }

    private func findFlight(_ bearer: String, _ callsign: String, _ origin: Airport, _ destination: Airport,
                            _ around: Date, _ durationMinutes: Int, _ seen: inout [String]) async throws -> (String, Int, Int)? {
        let prefix = String(callsign.prefix { $0.isLetter })
        func pick(_ body: Data?) -> (String, Int, Int)? {
            guard let body, let entries = try? JSONSerialization.jsonObject(with: body) as? [[String: Any]] else { return nil }
            for e in entries {
                let cs = (e["callsign"] as? String ?? "").trimmingCharacters(in: .whitespaces)
                if !cs.isEmpty, cs.hasPrefix(prefix), !seen.contains(cs) { seen.append(cs) }
            }
            guard let m = entries.first(where: { Self.sameCallsign($0["callsign"] as? String ?? "", callsign) }),
                  let icao = m["icao24"] as? String, let first = m["firstSeen"] as? Int, let last = m["lastSeen"] as? Int else { return nil }
            return (icao, first, last)
        }
        let begin = Int(around.timeIntervalSince1970) - 3 * 3600
        let end = Int(around.timeIntervalSince1970) + 6 * 3600
        if let r = pick(try await getJSON("\(Self.api)/flights/departure?airport=\(origin.icao)&begin=\(begin)&end=\(end)", bearer)) { return r }
        let arrivalEnd = end + durationMinutes * 60 + 3600
        if let r = pick(try await getJSON("\(Self.api)/flights/arrival?airport=\(destination.icao)&begin=\(begin)&end=\(arrivalEnd)", bearer)) { return r }
        for offset in [-1, 1] {
            let b = Int(around.timeIntervalSince1970) + offset * 3600
            if let r = pick(try await getJSON("\(Self.api)/flights/all?begin=\(b)&end=\(b + 2 * 3600)", bearer)) { return r }
        }
        return nil
    }

    private func findAirborne(_ bearer: String, _ callsign: String, _ origin: Airport, _ destination: Airport) async throws -> String? {
        let pad = 4.0
        let url = "\(Self.api)/states/all?lamin=\(min(origin.latitude, destination.latitude) - pad)&lomin=\(min(origin.longitude, destination.longitude) - pad)&lamax=\(max(origin.latitude, destination.latitude) + pad)&lomax=\(max(origin.longitude, destination.longitude) + pad)"
        guard let body = try await getJSON(url, bearer),
              let obj = try? JSONSerialization.jsonObject(with: body) as? [String: Any],
              let states = obj["states"] as? [[Any]] else { return nil }
        for v in states where v.count > 1 {
            if Self.sameCallsign(v[1] as? String ?? "", callsign) { return v[0] as? String }
        }
        return nil
    }

    /// OpenSky pads callsigns to eight characters and some carriers zero-pad the number.
    private static func sameCallsign(_ seen: String, _ wanted: String) -> Bool {
        func norm(_ s: String) -> String {
            let t = s.trimmingCharacters(in: .whitespaces).uppercased()
            let letters = t.prefix { $0.isLetter }
            let digits = t.dropFirst(letters.count).prefix { $0.isNumber }
            return String(letters) + String(digits.drop { $0 == "0" })
        }
        let w = norm(wanted)
        return !w.isEmpty && norm(seen) == w
    }

    private func fetchPath(_ bearer: String, _ icao24: String, at: Int) async throws -> [TrackPoint] {
        guard let body = try await getJSON("\(Self.api)/tracks/all?icao24=\(icao24)&time=\(at)", bearer) else {
            throw OpenSkyError(message: "OpenSky matched the aircraft (\(icao24)) but kept no track for that flight.")
        }
        guard let obj = try? JSONSerialization.jsonObject(with: body) as? [String: Any], let path = obj["path"] as? [[Any]] else {
            throw OpenSkyError(message: "OpenSky returned a track with no positions.")
        }
        let points = path.compactMap { p -> TrackPoint? in
            guard p.count > 2, let lat = p[1] as? Double, let lon = p[2] as? Double else { return nil }
            return TrackPoint(lat: lat, lon: lon)
        }
        if points.count < 2 { throw OpenSkyError(message: "Track too short to draw.") }
        return points
    }

    private func accessToken() async throws -> String {
        if let token, Date() < tokenExpiresAt { return token }
        guard Config.isOpenSkyConfigured else { throw OpenSkyError(message: "OpenSky client credentials are not configured.") }
        var req = URLRequest(url: Self.tokenURL)
        req.httpMethod = "POST"
        req.setValue("application/x-www-form-urlencoded", forHTTPHeaderField: "Content-Type")
        let form = "grant_type=client_credentials&client_id=\(Config.openSkyClientId)&client_secret=\(Config.openSkyClientSecret)"
        req.httpBody = Data(form.utf8)
        let (data, resp) = try await URLSession.shared.data(for: req)
        guard (resp as? HTTPURLResponse)?.statusCode == 200,
              let obj = try? JSONSerialization.jsonObject(with: data) as? [String: Any],
              let t = obj["access_token"] as? String else {
            throw OpenSkyError(message: "OpenSky refused the client credentials.")
        }
        token = t
        tokenExpiresAt = Date().addingTimeInterval(TimeInterval((obj["expires_in"] as? Int ?? 1800) - 60))
        return t
    }

    /// nil on 404 (OpenSky's "nothing here"); an error for anything else.
    private func getJSON(_ url: String, _ bearer: String) async throws -> Data? {
        var req = URLRequest(url: URL(string: url)!, timeoutInterval: 30)
        req.setValue("Bearer \(bearer)", forHTTPHeaderField: "Authorization")
        let (data, resp) = try await URLSession.shared.data(for: req)
        switch (resp as? HTTPURLResponse)?.statusCode ?? 0 {
        case 200: return data
        case 404: return nil
        case 429: throw OpenSkyError(message: "OpenSky rate limit reached for today.")
        case let c: throw OpenSkyError(message: "OpenSky answered HTTP \(c).")
        }
    }
}
